vault/external_tests/sealmigration/seal_migration_test.go

```diff
@@ -175,6 +186,258 @@ func migrateFromShamirToTransit_Pre14(
+		// equivalent of doing the following in HCL:
+		//     seal "transit" {
+		//       // ...
+		//       disabled = "true"
+		//     }
+		tcc.CoreConfig.UnwrapSeal = transitSeal
+	}
+
+	// Restart each follower with the new config, and migrate to Shamir.
+	leaderIdx := migratePost14(
+		t, logger, storage, cluster, opts,
+		cluster.RootToken, cluster.RecoveryKeys,
+		migrateTransitToShamir, modifyCoreConfig)
+	leader := cluster.Cores[leaderIdx]
+
+	// Read the secret
+	secret, err := leader.Client.Logical().Read("secret/foo")
+	if err != nil {
+		t.Fatal(err)
+	}
+	if diff := deep.Equal(secret.Data, map[string]interface{}{"zork": "quux"}); len(diff) > 0 {
+		t.Fatal(diff)
+	}
+
+	// Make sure the seal configs were updated correctly.
+	b, r, err := cluster.Cores[0].Core.PhysicalSealConfigs(context.Background())
+	if err != nil {
+		t.Fatal(err)
+	}
+	verifyBarrierConfig(t, b, wrapping.Shamir, keyShares, keyThreshold, 1)
+	if r != nil {
+		t.Fatalf("expected nil recovery config, got: %#v", r)
+	}
+}
+
+type migrationDirection int
+
+const (
+	migrateShamirToTransit migrationDirection = iota
+	migrateTransitToShamir
+)
+
+func migratePost14(
+	t *testing.T, logger hclog.Logger,
+	storage teststorage.ReusableStorage,
+	cluster *vault.TestCluster, opts *vault.TestClusterOptions,
+	rootToken string, recoveryKeys [][]byte,
+	migrate migrationDirection,
+	modifyCoreConfig func(*vault.TestClusterCore),
+) int {
+
+	// Restart each follower with the new config, and migrate.
+	for i := 1; i < len(cluster.Cores); i++ {
+		cluster.StopCore(t, i)
+		if storage.IsRaft {
+			teststorage.CloseRaftStorage(t, cluster, i)
+		}
+		modifyCoreConfig(cluster.Cores[i])
+		cluster.RestartCore(t, i, opts)
+
+		cluster.Cores[i].Client.SetToken(rootToken)
+		unsealMigrate(t, cluster.Cores[i].Client, recoveryKeys, true)
+		time.Sleep(5 * time.Second)
+	}
+
+	// Bring down the leader
+	cluster.StopCore(t, 0)
+	if storage.IsRaft {
+		teststorage.CloseRaftStorage(t, cluster, 0)
+	}
+
+	// Wait for the followers to establish a new leader
+	leaderIdx, err := testhelpers.AwaitLeader(t, cluster)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if leaderIdx == 0 {
+		t.Fatalf("Core 0 cannot be the leader right now")
+	}
+	leader := cluster.Cores[leaderIdx]
+	leader.Client.SetToken(rootToken)
+
+	// Bring core 0 back up
+	cluster.RestartCore(t, 0, opts)
+	cluster.Cores[0].Client.SetToken(rootToken)
+
+	// TODO look into why this is different for different migration directions,
+	// and why it is swapped for raft.
+	switch migrate {
+	case migrateShamirToTransit:
+		if storage.IsRaft {
+			unsealMigrate(t, cluster.Cores[0].Client, recoveryKeys, true)
+		} else {
+			unseal(t, cluster.Cores[0].Client, recoveryKeys)
+		}
+	case migrateTransitToShamir:
+		if storage.IsRaft {
+			unseal(t, cluster.Cores[0].Client, recoveryKeys)
+		} else {
+			unsealMigrate(t, cluster.Cores[0].Client, recoveryKeys, true)
+		}
+	default:
+		t.Fatalf("unreachable")
+	}
+
+	time.Sleep(5 * time.Second)
```

Comment:
perhaps add a comment here on why the sleep is necessary? Can this result in any non-determinism?

Reply:
Nice catch -- this line turned out not to be necessary, so I removed it.